### FASE 1. Carga e integración de datos

**Proyecto:** Análisis Exploratorio de Datos — Ingresos totales por ventas a clientes de un comercio.

**Objetivo de esta fase:**
Importar los datos desde las distintas fuentes disponibles y preparar un único dataset consolidado sobre el que trabajar en las siguientes fases (limpieza, EDA e informe).

**Fuentes de datos:**
- `clientes.csv` → información de los clientes  (edad, estado civil, ingresos, ciudad, , etc.)
- `ventas.xlsx` → 5 hojas: 1 hoja con los productos y la información de las ventas, repartidas en 4 hojas (2022, 2023, 2024, 2025)


### 1️⃣ Importar librerías

In [1]:
import pandas as pd
import numpy as np

# Para que se muestren todas las columnas al inspeccionar los DataFrames
pd.set_option('display.max_columns', None)

In [ ]:
# Libreria de Tratamiento de Datos
import pandas as pd
from IPython.display import display

# Exploracion de Archivos
import sys
sys.path.append('../') # para que busque en carpeta src 
from src import sp_carga_integracion as ci

# Funcion personalizada para asegurarnos que aparezcan todas las columnas
pd.set_option('display.max_columns', None) 

### 2️⃣ Cargar `clientes.csv`
Este archivo contiene la informacion de los clientes.

In [2]:
df_clientes = pd.read_csv(r"../data/raw/clientes.csv")

#### Comprobación de la carga

In [3]:
# Primeras 5 filas
df_clientes.head()

,Unnamed: 0,id_,edad,estado_civil,nivel_educativo,ingresos_anuales,tiene_hijos,adolescentes_en_casa,fecha_registro,ciudad,region,latitude,longitude,segmento_cliente,visitas_web_mes
0,0,b9e805ce-fe13-4ed5-82b1-286271529391,44.0,casado,grado universitario,30544,1.0,0.0,20-octubre-2021,Barcelona,Este,39.838,-7.047,Recurrente,9
1,1,fa2c965b-9356-48b4-84f1-ee1bf213cee9,26.0,casado,bachillerato,46779,0.0,0.0,21-agosto-2019,Malaga,Sur,39.741,0.201,VIP,4
2,2,0dd76b11-d81d-4df7-8ef0-23dab7ccf86f,50.0,Casado,master,22003,1.0,0.0,3-diciembre-2020,Vigo,Norte,38.971,-6.919,Nuevo,8
3,3,ac5ab989-4b81-4b83-8b09-5842b10bacc4,52.0,casado,grado universitario,41512,0.0,0.0,29-diciembre-2021,Malaga,Sur,36.810,1.782,Nuevo,8
4,4,11f866be-75b7-40d2-a32f-2161caf21791,18.0,Casado,bachillerato,45029,0.0,0.0,4-septiembre-2021,Valencia,Este,39.879,-0.514,Corporativo,4


In [4]:
#  Dimensiones del dataset
print("(Filas , Columnas):",df_clientes.shape)

(Filas , Columnas): (8000, 15)


In [5]:
# Nombres de las columnas
print("El archivo df_clientes esta compuesto por las columnas:\n",df_clientes.columns)


El archivo df_clientes esta compuesto por las columnas:
 Index(['Unnamed: 0', 'id_', 'edad', 'estado_civil', 'nivel_educativo',
       'ingresos_anuales', 'tiene_hijos', 'adolescentes_en_casa',
       'fecha_registro', 'ciudad', 'region', 'latitude', 'longitude',
       'segmento_cliente', 'visitas_web_mes'],
      dtype='str')


In [6]:
# Tipos de datos y nulos por columnas
df_clientes.info()

<class 'pandas.DataFrame'>
RangeIndex: 8000 entries, 0 to 7999
Data columns (total 15 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Unnamed: 0            8000 non-null   int64  
 1   id_                   8000 non-null   str    
 2   edad                  7520 non-null   float64
 3   estado_civil          7920 non-null   str    
 4   nivel_educativo       7680 non-null   str    
 5   ingresos_anuales      8000 non-null   int64  
 6   tiene_hijos           7840 non-null   float64
 7   adolescentes_en_casa  8000 non-null   float64
 8   fecha_registro        7960 non-null   str    
 9   ciudad                8000 non-null   str    
 10  region                8000 non-null   str    
 11  latitude              8000 non-null   float64
 12  longitude             8000 non-null   float64
 13  segmento_cliente      8000 non-null   str    
 14  visitas_web_mes       8000 non-null   int64  
dtypes: float64(5), int64(3), str(7)


In [1]:
# Libreria de Tratamiento de Datos
import pandas as pd
from IPython.display import display

# Exploracion de Archivos
import sys
sys.path.append('../') # para que busque en carpeta src 
from src import sp_carga_integracion as ci 

# Funcion personalizada para asegurarnos que aparezcan todas las columnas
pd.set_option('display.max_columns', None) 

In [2]:
df_clientes = pd.read_csv(r"../data/raw/clientes.csv")

In [3]:
ci.eda_preliminar(df_clientes)

5 FILAS ALEATORIAS


,1865,1017,6128,3103,2155
Unnamed: 0,1865,1017,6128,3103,2155
id_,df89d3ca-d03b-4a45-9029-d8c501f86287,77aaff2c-616d-4bd4-b5c5-44fe12ea64dc,8a2c6d94-b7bf-4bac-bb95-528100ae1e41,603ecf88-9196-422e-8966-78a44ed06737,8fe9a3d2-a0b5-4241-b9fc-de1e52675c7e
edad,42.0,36.0,57.0,54.0,48.0
estado_civil,DIVORCIADO,Casado,SOLTERO,Casado,CASADO
nivel_educativo,grado universitario,NaN,master,grado universitario,bachillerato
ingresos_anuales,23705,52264,37103,64728,33413
tiene_hijos,0.0,1.0,1.0,0.0,0.0
adolescentes_en_casa,0.0,0.0,0.0,0.0,0.0
fecha_registro,28-febrero-2021,21-abril-2021,24-febrero-2021,31-enero-2020,20-junio-2021
ciudad,Zaragoza,Madrid,Murcia,Valencia,Barcelona


DIMENSIONES
El munero de filas es: 8000 filas 
El munero de columnas es: 15 columnas
INFORMACION BASICA
<class 'pandas.DataFrame'>
RangeIndex: 8000 entries, 0 to 7999
Data columns (total 15 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Unnamed: 0            8000 non-null   int64  
 1   id_                   8000 non-null   str    
 2   edad                  7520 non-null   float64
 3   estado_civil          7920 non-null   str    
 4   nivel_educativo       7680 non-null   str    
 5   ingresos_anuales      8000 non-null   int64  
 6   tiene_hijos           7840 non-null   float64
 7   adolescentes_en_casa  8000 non-null   float64
 8   fecha_registro        7960 non-null   str    
 9   ciudad                8000 non-null   str    
 10  region                8000 non-null   str    
 11  latitude              8000 non-null   float64
 12  longitude             8000 non-null   float64
 13  segmento_cliente      8000 non

,count,unique,top,freq
id_,8000,8000,b9e805ce-fe13-4ed5-82b1-286271529391,1
estado_civil,7920,9,Casado,1337
nivel_educativo,7680,5,grado universitario,2703
fecha_registro,7960,1802,7-marzo-2019,12
ciudad,8000,10,Bilbao,855
region,8000,4,Sur,3264
segmento_cliente,8000,4,Recurrente,3195


ESTADISTICOS NUMÉRICAS


,count,mean,std,min,25%,50%,75%,max
Unnamed: 0,8000.0,3999.500000,2309.545410,0.000,1999.750,3999.5000,5999.2500,7999.000
edad,7520.0,40.188298,12.502892,18.000,31.000,40.0000,48.0000,85.000
ingresos_anuales,8000.0,43943.856375,16109.038812,14057.000,32114.250,41214.5000,52988.5000,149422.000
tiene_hijos,7840.0,0.446556,0.497167,0.000,0.000,0.0000,1.0000,1.000
adolescentes_en_casa,8000.0,0.306875,0.461226,0.000,0.000,0.0000,1.0000,1.000
latitude,8000.0,39.771402,2.153900,36.001,37.918,39.7945,41.6050,43.499
longitude,8000.0,-2.830366,3.524089,-8.999,-5.853,-2.8115,0.1825,3.299
visitas_web_mes,8000.0,8.018625,2.809851,0.000,6.000,8.0000,10.0000,21.000


**Observaciones iniciales de `df_clientes`:**
- Aparece una columna `Unnamed: 0` que es un índice residual del archivo original y no aporta información → se eliminará.
- La columna `fecha_registro` se han cargado como texto (`str`). Esto se corregirá en la **Fase 2 (Limpieza)**.
- Hay columnas con valores nulos (`edad`, `estado_civil`, `nivel_educativo`, `tiene_hijos`, `fecha_registro`),se tratarán en la **Fase 2 (Limpieza)**.
- Las columnas (`tiene_hijos`, `adolescentes_en_casa`) en una primera visual parecen variables binarias, se tratarán en la **Fase 2 (Limpieza)**.


#### 3️⃣ Cargar `ventas.csv`

Este archivo contiene 5 hojas, la HOJA 1 con los productos dispoblibles en tienda, y las HOJAS 2 A 5 con las ventas realizadas en los años (2022, 2023, 2024, 2025).


In [7]:
# Comprobamos las hojas que tiene el Excel
excel_ventas = pd.ExcelFile(r"../data/raw/ventas.xlsx")
print("El excel ventas lo compone las Hojas:\n",excel_ventas.sheet_names)

El excel ventas lo compone las Hojas:
 ['productos', '2022', '2023', '2024', '2025']


#### **Hoja Productos**

In [8]:
# Cargamos la hoja productos
df_productos = pd.read_excel(excel_ventas, sheet_name="productos")

In [9]:
# Primeras 5 filas
df_productos.head()

,Unnamed: 0,product_id,categoria_producto,nombre_producto,marca,costo_unitario,precio_lista,peso_kg,proveedor,stock_disponible,fecha_lanzamiento,rating_promedio_producto
0,0,P0001,Electronica,Auriculares Urbanix Basic,Urbanix,"26,73","45,70",9.60,Global Supply SL,383,31-mayo-2020,3.60
1,1,P0002,Electronica,Auriculares Urbanix Pro,Urbanix,"47,97","77,93",11.86,Global Supply SL,392,8-marzo-2019,3.29
2,2,P0003,Electronica,Auriculares Nova Max,Nova,"49,49","98,89",8.54,Global Supply SL,317,16-junio-2023,4.18
3,3,P0004,Electronica,Smartphone Norte&Co Basic,Norte&Co,"60,57","95,49",3.60,Comercial Atlas,303,23-enero-2019,2.94
4,4,P0005,Electronica,Smartphone Zentra Pro,Zentra,"17,49","26,00",1.69,Norte Logistica,387,30-enero-2021,3.95


In [10]:
#  Dimensiones del dataset
print("(Filas , Columnas):",df_productos.shape)

(Filas , Columnas): (90, 12)


In [11]:
# Nombres de las columnas
print("El archivo df_clientes esta compuesto por las columnas:\n",df_productos.columns)

El archivo df_clientes esta compuesto por las columnas:
 Index(['Unnamed: 0', 'product_id', 'categoria_producto', 'nombre_producto',
       'marca', 'costo_unitario', 'precio_lista', 'peso_kg', 'proveedor',
       'stock_disponible', 'fecha_lanzamiento', 'rating_promedio_producto'],
      dtype='str')


In [12]:
# Tipos de datos y nulos por columnas
df_productos.info()

<class 'pandas.DataFrame'>
RangeIndex: 90 entries, 0 to 89
Data columns (total 12 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Unnamed: 0                90 non-null     int64  
 1   product_id                90 non-null     str    
 2   categoria_producto        90 non-null     str    
 3   nombre_producto           90 non-null     str    
 4   marca                     89 non-null     str    
 5   costo_unitario            90 non-null     str    
 6   precio_lista              90 non-null     str    
 7   peso_kg                   90 non-null     float64
 8   proveedor                 90 non-null     str    
 9   stock_disponible          90 non-null     int64  
 10  fecha_lanzamiento         90 non-null     str    
 11  rating_promedio_producto  88 non-null     float64
dtypes: float64(2), int64(2), str(8)
memory usage: 15.4 KB


**Observaciones iniciales de `df_productos`:**
- Aparece una columna `Unnamed: 0` que es un índice residual del archivo original y no aporta información → se eliminará.
- La columna `fecha_lanzamiento` se ha cargado como texto (`str`).Se corregirá en la **Fase 2 (Limpieza)**.
- Hay dos columna con valores nulos (`rating_promedio_producto`, `marca`),se tratarán en la **Fase 2 (Limpieza)**.
- Las columnas (`costo_unitario`, `precio_lista`) se cargaron como texto (`str`), se tratarán en la **Fase 2 (Limpieza)**.
- Se observa que la `columna product_id` esta relacionada con las Hojas 2022, 2023, 2024 ,2025, se comprobara antes de unirlas. 


#### **Hojas 2022, 2023, 2024 ,2025**

In [13]:
# Cargamos cada hoja por separado
df_2022 = pd.read_excel(excel_ventas,sheet_name="2022")
df_2023 = pd.read_excel(excel_ventas,sheet_name="2023")
df_2024 = pd.read_excel(excel_ventas,sheet_name="2024")
df_2025 = pd.read_excel(excel_ventas,sheet_name="2025")

In [14]:
#  Dimensiones del dataset
for año, df in [("2022", df_2022), ("2023", df_2023), ("2024", df_2024), ("2025", df_2025)]:
    print(f"Año {año}: tiene {df.shape[0]} Filas y {df.shape[1]} Columnas ")
    

Año 2022: tiene 13000 Filas y 14 Columnas 
Año 2023: tiene 13000 Filas y 14 Columnas 
Año 2024: tiene 13000 Filas y 14 Columnas 
Año 2025: tiene 13000 Filas y 14 Columnas 


In [15]:
# Nombres de las columnas
for año, df in [("2022", df_2022), ("2023", df_2023), ("2024", df_2024), ("2025", df_2025)]:
    print(f"Columnas de {año}: {df.columns.tolist()}")

Columnas de 2022: ['Unnamed: 0', 'order_id', 'id_', 'product_id', 'fecha_pedido', 'cantidad', 'precio_unitario', 'descuento', 'importe_total', 'metodo_pago', 'envio_gratis', 'calificacion', 'devuelto', 'canal']
Columnas de 2023: ['Unnamed: 0', 'order_id', 'id_', 'product_id', 'fecha_pedido', 'cantidad', 'precio_unitario', 'descuento', 'importe_total', 'metodo_pago', 'envio_gratis', 'calificacion', 'devuelto', 'canal']
Columnas de 2024: ['Unnamed: 0', 'order_id', 'id_', 'product_id', 'fecha_pedido', 'cantidad', 'precio_unitario', 'descuento', 'importe_total', 'metodo_pago', 'envio_gratis', 'calificacion', 'devuelto', 'canal']
Columnas de 2025: ['Unnamed: 0', 'order_id', 'id_', 'product_id', 'fecha_pedido', 'cantidad', 'precio_unitario', 'descuento', 'importe_total', 'metodo_pago', 'envio_gratis', 'calificacion', 'devuelto', 'canal']


In [16]:
df_2022.sample(3)

,Unnamed: 0,order_id,id_,product_id,fecha_pedido,cantidad,precio_unitario,descuento,importe_total,metodo_pago,envio_gratis,calificacion,devuelto,canal
9302,9302,9303,c1a898b1-45c4-4ebb-9016-4e10ec3dc7ea,P0025,11-agosto-2022,3,"44,62",0.10,"120,46",Tarjeta de credito,1,5.0,no,App movil
43,43,44,f0217079-a404-4bf0-9048-161e5857f86c,P0084,23-marzo-2022,1,"890,29",0.10,"801,26",TARJETA DE CREDITO,0,4.0,no,App movil
4972,4972,4973,87ceaaab-25fa-4557-a501-bd214678440d,P0084,16-octubre-2022,2,"53,06",0.15,"90,21",Tarjeta de debito,0,5.0,NO,Marketplace


In [17]:
df_2023.sample(3)

,Unnamed: 0,order_id,id_,product_id,fecha_pedido,cantidad,precio_unitario,descuento,importe_total,metodo_pago,envio_gratis,calificacion,devuelto,canal
2997,2997,15998,01281c4d-ed60-4c8a-bd69-e75769863a39,P0037,2-septiembre-2023,1,"36,51",0.00,"36,51",CONTRAREEMBOLSO,1,4.0,no,Tienda fisica
12347,12347,25348,51497e32-1b9c-4a1d-bb82-d9ed29983a73,P0038,7-marzo-2023,1,"50,39",0.05,"47,87",tarjeta de debito,0,4.0,no,Marketplace
7107,7107,20108,0a8221af-4734-4860-a4e0-9ad05718ddf1,P0083,16-enero-2023,2,"80,72",0.10,"145,30",Tarjeta de credito,0,NaN,NO,App movil


In [18]:
df_2024.sample(3)

,Unnamed: 0,order_id,id_,product_id,fecha_pedido,cantidad,precio_unitario,descuento,importe_total,metodo_pago,envio_gratis,calificacion,devuelto,canal
5314,5314,31315,cf811157-d518-44e9-975d-b867c8df9d36,P0030,4-abril-2024,2,"41,15",0.05,"78,19",TARJETA DE CREDITO,1,5.0,no,Web
791,791,26792,9a626669-3c40-4cb5-bf5c-e39cd6e7e60d,P0042,15-noviembre-2024,1,"63,51",0.15,"53,98",CONTRAREEMBOLSO,0,4.0,no,App movil
2297,2297,28298,1ba188c0-529d-4395-b711-e6fed0119515,P0082,24-enero-2024,4,"48,36",0.00,"193,44",contrareembolso,0,NaN,no,Tienda fisica


In [19]:
df_2025.sample(3)

,Unnamed: 0,order_id,id_,product_id,fecha_pedido,cantidad,precio_unitario,descuento,importe_total,metodo_pago,envio_gratis,calificacion,devuelto,canal
4411,4411,43412,ac09f775-f0e2-4622-b39f-a6870eecda6f,P0066,18-octubre-2025,3,"45,41",0.0,"136,24",TARJETA DE CREDITO,1,5.0,NO,Tienda fisica
4012,4012,43013,8c6c2cb1-d76d-49f3-887c-2192aa7d782a,P0085,10-agosto-2025,5,"42,72",0.0,"213,59",tarjeta de credito,0,NaN,no,Web
11751,11751,50752,81278a1a-b3dc-4468-92de-e7eeb3140ee4,P0090,13-julio-2025,1,"28,46",0.0,"28,46",PAYPAL,0,5.0,no,Web


In [20]:
df_2022.info()

<class 'pandas.DataFrame'>
RangeIndex: 13000 entries, 0 to 12999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Unnamed: 0       13000 non-null  int64  
 1   order_id         13000 non-null  int64  
 2   id_              13000 non-null  str    
 3   product_id       13000 non-null  str    
 4   fecha_pedido     13000 non-null  str    
 5   cantidad         13000 non-null  int64  
 6   precio_unitario  13000 non-null  str    
 7   descuento        13000 non-null  float64
 8   importe_total    13000 non-null  str    
 9   metodo_pago      12805 non-null  str    
 10  envio_gratis     13000 non-null  int64  
 11  calificacion     11125 non-null  float64
 12  devuelto         13000 non-null  str    
 13  canal            13000 non-null  str    
dtypes: float64(2), int64(4), str(8)
memory usage: 2.5 MB


In [21]:
df_2023.info()

<class 'pandas.DataFrame'>
RangeIndex: 13000 entries, 0 to 12999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Unnamed: 0       13000 non-null  int64  
 1   order_id         13000 non-null  int64  
 2   id_              13000 non-null  str    
 3   product_id       13000 non-null  str    
 4   fecha_pedido     13000 non-null  str    
 5   cantidad         13000 non-null  int64  
 6   precio_unitario  13000 non-null  str    
 7   descuento        13000 non-null  float64
 8   importe_total    13000 non-null  str    
 9   metodo_pago      12805 non-null  str    
 10  envio_gratis     13000 non-null  int64  
 11  calificacion     11072 non-null  float64
 12  devuelto         13000 non-null  str    
 13  canal            13000 non-null  str    
dtypes: float64(2), int64(4), str(8)
memory usage: 2.5 MB


In [22]:
df_2024.info()

<class 'pandas.DataFrame'>
RangeIndex: 13000 entries, 0 to 12999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Unnamed: 0       13000 non-null  int64  
 1   order_id         13000 non-null  int64  
 2   id_              13000 non-null  str    
 3   product_id       13000 non-null  str    
 4   fecha_pedido     13000 non-null  str    
 5   cantidad         13000 non-null  int64  
 6   precio_unitario  13000 non-null  str    
 7   descuento        13000 non-null  float64
 8   importe_total    13000 non-null  str    
 9   metodo_pago      12805 non-null  str    
 10  envio_gratis     13000 non-null  int64  
 11  calificacion     11053 non-null  float64
 12  devuelto         13000 non-null  str    
 13  canal            13000 non-null  str    
dtypes: float64(2), int64(4), str(8)
memory usage: 2.5 MB


In [23]:
df_2025.info()

<class 'pandas.DataFrame'>
RangeIndex: 13000 entries, 0 to 12999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Unnamed: 0       13000 non-null  int64  
 1   order_id         13000 non-null  int64  
 2   id_              13000 non-null  str    
 3   product_id       13000 non-null  str    
 4   fecha_pedido     13000 non-null  str    
 5   cantidad         13000 non-null  int64  
 6   precio_unitario  13000 non-null  str    
 7   descuento        13000 non-null  float64
 8   importe_total    13000 non-null  str    
 9   metodo_pago      12805 non-null  str    
 10  envio_gratis     13000 non-null  int64  
 11  calificacion     10977 non-null  float64
 12  devuelto         13000 non-null  str    
 13  canal            13000 non-null  str    
dtypes: float64(2), int64(4), str(8)
memory usage: 2.5 MB


**Observaciones iniciales de `df_2022, df_2023, df_2024, df_2025`:**
- Aparece una columna `Unnamed: 0` que es un índice residual del archivo original y no aporta información → se eliminará.
- La columna `fecha_pedido` se ha cargado como texto (`str`).Se corregirá en la **Fase 2**.
- Hay dos columnas con valores nulos (`metodo_pago`,`calificacion` ),se tratarán en la **Fase 2**.
- Las columnas (`precio_unitario`, `importe_total`) se cargaron como texto (`str`), se tratarán en la **Fase 2**.
- La columna (`envio_gratis`) en una primera visual parece una variable binaria, se tratarán en la **Fase 2**.
- Se observa que la `columna product_id` esta relacionada con las Hojas 2022, 2023, 2024 ,2025, se comprobara antes de unirlas. 


#### 4️⃣ Unir  `df_2022, df_2023, df_2024, df_2025` del archivo `ventas.csv`

Como las 4 hojas tienen exactamente las mismas columnas y representan el mismo tipo de información, las unimos verticalmente con pd.concat().

Antes de unirlas, añadimos una columna `hoja_origen` para no perder la trazabilidad de qué hoja proviene cada registro (nos sera útil más adelante).


In [24]:
# Añadimos una columna que identifique el año/hoja de origen antes de concatenar
df_2022["hoja_origen"] = "2022"
df_2023["hoja_origen"] = "2023"
df_2024["hoja_origen"] = "2024"
df_2025["hoja_origen"] = "2025"


In [25]:
# Verificamos que se ha añadido la columna "hoja_origen"
#  Dimensiones del dataset
for año, df in [("2022", df_2022), ("2023", df_2023), ("2024", df_2024), ("2025", df_2025)]:
    print(f"Año {año}: tiene {df.shape[0]} Filas y {df.shape[1]} Columnas ")
    

Año 2022: tiene 13000 Filas y 15 Columnas 
Año 2023: tiene 13000 Filas y 15 Columnas 
Año 2024: tiene 13000 Filas y 15 Columnas 
Año 2025: tiene 13000 Filas y 15 Columnas 


In [26]:
# Nombres de las columnas
for año, df in [("2022", df_2022), ("2023", df_2023), ("2024", df_2024), ("2025", df_2025)]:
    print(f"Columnas de {año}: {df.columns.tolist()}")

Columnas de 2022: ['Unnamed: 0', 'order_id', 'id_', 'product_id', 'fecha_pedido', 'cantidad', 'precio_unitario', 'descuento', 'importe_total', 'metodo_pago', 'envio_gratis', 'calificacion', 'devuelto', 'canal', 'hoja_origen']
Columnas de 2023: ['Unnamed: 0', 'order_id', 'id_', 'product_id', 'fecha_pedido', 'cantidad', 'precio_unitario', 'descuento', 'importe_total', 'metodo_pago', 'envio_gratis', 'calificacion', 'devuelto', 'canal', 'hoja_origen']
Columnas de 2024: ['Unnamed: 0', 'order_id', 'id_', 'product_id', 'fecha_pedido', 'cantidad', 'precio_unitario', 'descuento', 'importe_total', 'metodo_pago', 'envio_gratis', 'calificacion', 'devuelto', 'canal', 'hoja_origen']
Columnas de 2025: ['Unnamed: 0', 'order_id', 'id_', 'product_id', 'fecha_pedido', 'cantidad', 'precio_unitario', 'descuento', 'importe_total', 'metodo_pago', 'envio_gratis', 'calificacion', 'devuelto', 'canal', 'hoja_origen']


#### Concatenamos 

In [27]:
df_ventas = pd.concat([df_2022, df_2023, df_2024,df_2025], ignore_index=True)

In [28]:
df_ventas.sample(5)

,Unnamed: 0,order_id,id_,product_id,fecha_pedido,cantidad,precio_unitario,descuento,importe_total,metodo_pago,envio_gratis,calificacion,devuelto,canal,hoja_origen
32952,6952,32953,416a14ab-0db9-40cf-bbea-03f6ce8c37e8,P0071,14-enero-2024,1,"41,69",0.10,"37,52",tarjeta de credito,1,5.0,NO,App movil,2024
7356,7356,7357,f78b6ef8-ed82-4353-b6d1-4ac0860af7f9,P0076,29-marzo-2022,2,"22,48",0.20,"35,96",Tarjeta de credito,0,4.0,no,App movil,2022
33350,7350,33351,027b66e4-055e-4268-a62f-2440676cec85,P0076,16-febrero-2024,1,"31,95",0.15,"27,16",TARJETA DE CREDITO,1,3.0,no,Tienda fisica,2024
21533,8533,21534,a0df0a1a-57e2-40bd-af50-b49f21198d3d,P0075,12-septiembre-2023,1,"60,12",0.05,"57,11",Tarjeta de debito,0,5.0,no,Web,2023
25304,12304,25305,4d58d282-62ee-4066-b5ec-336fd401bada,P0033,8-julio-2023,1,"74,49",0.00,"74,49",TARJETA DE CREDITO,0,4.0,NO,App movil,2023


#### Comprobación de la concatenación

In [43]:
# La suma de filas debe coincidir con la suma de las 3 hojas por separado
print("Filas 2022 + 2023 + 2024 + 2025:", len(df_2022) + len(df_2023) + len(df_2024) + len(df_2025))
print("Filas df_ventas:", len(df_ventas))

Filas 2022 + 2023 + 2024 + 2025: 52000
Filas df_ventas: 52000


In [30]:
#  Dimensiones del dataset
print("(Filas , Columnas):",df_ventas.shape)

(Filas , Columnas): (52000, 15)


In [31]:
# Reviamos el df_ventas 
df_ventas.info()

<class 'pandas.DataFrame'>
RangeIndex: 52000 entries, 0 to 51999
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Unnamed: 0       52000 non-null  int64  
 1   order_id         52000 non-null  int64  
 2   id_              52000 non-null  str    
 3   product_id       52000 non-null  str    
 4   fecha_pedido     52000 non-null  str    
 5   cantidad         52000 non-null  int64  
 6   precio_unitario  52000 non-null  str    
 7   descuento        52000 non-null  float64
 8   importe_total    52000 non-null  str    
 9   metodo_pago      51220 non-null  str    
 10  envio_gratis     52000 non-null  int64  
 11  calificacion     44227 non-null  float64
 12  devuelto         52000 non-null  str    
 13  canal            52000 non-null  str    
 14  hoja_origen      52000 non-null  str    
dtypes: float64(2), int64(4), str(9)
memory usage: 10.6 MB


**Observaciones iniciales de `df_ventas`:**
- Aparece una columna `Unnamed: 0` que es un índice residual del archivo original y no aporta información → se eliminará.
- La columna `fecha_pedido` se ha cargado como texto (`str`).Se corregirá en la **Fase 2**.
- Hay dos columnas con valores nulos (`metodo_pago`,`calificacion` ),se tratarán en la **Fase 2**.
- Las columnas (`precio_unitario`, `importe_total`) se cargaron como texto (`str`), se tratarán en la **Fase 2**.0
- La columna (`envio_gratis`) en una primera visual parece una variable binaria, se tratarán en la **Fase 2**.
- Se observa que la `columna product_id` esta relacionada con las Hojas 2022, 2023, 2024 ,2025, se comprobara antes de unirlas. 


#### 5️⃣ Unir `df_ventas` con `df_productos`

Como hemos observado ambos comparten la columna `producto_id`, por lo que uniremos con `merge()` y con `how="left"` para conservar todas las dilas de `df_ventas`

In [32]:
# Comprobamos si hay algún product_id en df_ventas que no exista en df_productos
set(df_ventas["product_id"]) - set(df_productos["product_id"])

set()

In [33]:
df_ventas_productos = df_ventas.merge(df_productos, on="product_id", how="left")

In [42]:
print(len(df_ventas), len(df_ventas_productos))  # deberían ser iguales
df_ventas_productos.isnull().sum()  # si aparecen nulos en columnas de producto, algo no casó bien

52000 52000


Unnamed: 0_x                   0
order_id                       0
id_                            0
product_id                     0
fecha_pedido                   0
cantidad                       0
precio_unitario                0
descuento                      0
importe_total                  0
metodo_pago                  780
envio_gratis                   0
calificacion                7773
devuelto                       0
canal                          0
hoja_origen                    0
Unnamed: 0_y                   0
categoria_producto             0
nombre_producto                0
marca                        567
costo_unitario                 0
precio_lista                   0
peso_kg                        0
proveedor                      0
stock_disponible               0
fecha_lanzamiento              0
rating_promedio_producto    1133
dtype: int64

#### 6️⃣ Unimos df_ventas_productos con df_clientes

Como hemos observado ambos comparten la columna `id_`, por lo que uniremos con `merge()` y con `how="left"` para conservar todas las dilas de `df_ventas_productos`

In [34]:
df_clientes.info()

<class 'pandas.DataFrame'>
RangeIndex: 8000 entries, 0 to 7999
Data columns (total 15 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Unnamed: 0            8000 non-null   int64  
 1   id_                   8000 non-null   str    
 2   edad                  7520 non-null   float64
 3   estado_civil          7920 non-null   str    
 4   nivel_educativo       7680 non-null   str    
 5   ingresos_anuales      8000 non-null   int64  
 6   tiene_hijos           7840 non-null   float64
 7   adolescentes_en_casa  8000 non-null   float64
 8   fecha_registro        7960 non-null   str    
 9   ciudad                8000 non-null   str    
 10  region                8000 non-null   str    
 11  latitude              8000 non-null   float64
 12  longitude             8000 non-null   float64
 13  segmento_cliente      8000 non-null   str    
 14  visitas_web_mes       8000 non-null   int64  
dtypes: float64(5), int64(3), str(7)


In [35]:
df_ventas_productos.info()

<class 'pandas.DataFrame'>
RangeIndex: 52000 entries, 0 to 51999
Data columns (total 26 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Unnamed: 0_x              52000 non-null  int64  
 1   order_id                  52000 non-null  int64  
 2   id_                       52000 non-null  str    
 3   product_id                52000 non-null  str    
 4   fecha_pedido              52000 non-null  str    
 5   cantidad                  52000 non-null  int64  
 6   precio_unitario           52000 non-null  str    
 7   descuento                 52000 non-null  float64
 8   importe_total             52000 non-null  str    
 9   metodo_pago               51220 non-null  str    
 10  envio_gratis              52000 non-null  int64  
 11  calificacion              44227 non-null  float64
 12  devuelto                  52000 non-null  str    
 13  canal                     52000 non-null  str    
 14  hoja_origen      

In [36]:
# Unir con df_clientes por la columna id_
df_final = df_ventas_productos.merge(df_clientes, on="id_", how="left")

In [37]:
df_final.columns

Index(['Unnamed: 0_x', 'order_id', 'id_', 'product_id', 'fecha_pedido',
       'cantidad', 'precio_unitario', 'descuento', 'importe_total',
       'metodo_pago', 'envio_gratis', 'calificacion', 'devuelto', 'canal',
       'hoja_origen', 'Unnamed: 0_y', 'categoria_producto', 'nombre_producto',
       'marca', 'costo_unitario', 'precio_lista', 'peso_kg', 'proveedor',
       'stock_disponible', 'fecha_lanzamiento', 'rating_promedio_producto',
       'Unnamed: 0', 'edad', 'estado_civil', 'nivel_educativo',
       'ingresos_anuales', 'tiene_hijos', 'adolescentes_en_casa',
       'fecha_registro', 'ciudad', 'region', 'latitude', 'longitude',
       'segmento_cliente', 'visitas_web_mes'],
      dtype='str')

In [38]:
# Eliminamos las columnas Unnamed
df_final= df_final.drop(columns=["Unnamed: 0_x", "Unnamed: 0_y","Unnamed: 0"])

In [39]:
# comprobamos 
df_final.columns

Index(['order_id', 'id_', 'product_id', 'fecha_pedido', 'cantidad',
       'precio_unitario', 'descuento', 'importe_total', 'metodo_pago',
       'envio_gratis', 'calificacion', 'devuelto', 'canal', 'hoja_origen',
       'categoria_producto', 'nombre_producto', 'marca', 'costo_unitario',
       'precio_lista', 'peso_kg', 'proveedor', 'stock_disponible',
       'fecha_lanzamiento', 'rating_promedio_producto', 'edad', 'estado_civil',
       'nivel_educativo', 'ingresos_anuales', 'tiene_hijos',
       'adolescentes_en_casa', 'fecha_registro', 'ciudad', 'region',
       'latitude', 'longitude', 'segmento_cliente', 'visitas_web_mes'],
      dtype='str')

In [40]:
df_final.shape

(52000, 37)

In [41]:
print("Numero de filas en df_clientes:", len(df_clientes))
print("numero de filas en df_ventas_productos es:", len(df_ventas_productos))
print("numero de filas en df_final es :", len(df_final))


Numero de filas en df_clientes: 8000
numero de filas en df_ventas_productos es: 52000
numero de filas en df_final es : 52000


In [72]:
df_final.head(3).T

,0,1,2
order_id,1,2,3
id_,7ec28fa9-cad8-4f9a-a5b4-04a3f722c107,0f14d6ee-e853-448f-a20f-d6c933956321,5db0a392-f9b6-48c4-a1b0-bdd040ab4917
product_id,P0086,P0086,P0001
fecha_pedido,3-septiembre-2022,18-noviembre-2022,14-enero-2022
cantidad,4,1,1
precio_unitario,"27,29","68,24","25,84"
descuento,0.1,0.05,0.1
importe_total,"98,23","64,83","23,26"
metodo_pago,PAYPAL,PayPal,Tarjeta de credito
envio_gratis,0,0,0


In [73]:
# Comprobamos cuántos registros de la campaña son valores nulos, solo los registros con valores nulos  
nulos = df_final.isna().sum()
nulos[nulos>0].sort_values(ascending=False)


calificacion                7773
edad                        3039
nivel_educativo             2063
rating_promedio_producto    1133
tiene_hijos                 1038
metodo_pago                  780
marca                        567
estado_civil                 524
fecha_registro               253
dtype: int64

In [44]:
nulos_prod = df_productos.isna().sum()
nulos_prod[nulos_prod > 0].sort_values(ascending=True)

marca                       1
rating_promedio_producto    2
dtype: int64

**Observaciones iniciales de `df_final`:**
- Hay dos columnas con valores nulos (`marca`,`rating_promedio_producto`), aparecen 567 nulos en `marca` y 1133 nulos en `rating_promedio_producto` que no aparecian en df_productos, esto es debido a que se ha aumentado de 90 filas de `productos` a 52.000 filas de `ventas`, por lo que cada nulo de `productos` se ha repetido tantas veces como `ventas` tenga ese producto, se tratarán en la **Fase 2**.

### 7️⃣ Guardar el dataset integrado
Guardamos el resultado de esta fase como `01_datos_integrados` para poder continuar con la Fase 2 (Limpieza y transformación) sin tener que repetir la carga y el merge cada vez.


In [45]:
df_final.to_csv("01_datos_integrados.csv", index=False)